connectors/src/connectors/notion/lib/parents.ts

```diff
@@ -0,0 +1,142 @@
+import memoize from "lodash.memoize";
+
+import { updateDocumentParentsField } from "@connectors/lib/data_sources";
+import { NotionDatabase, NotionPage } from "@connectors/lib/models";
+import {
+  DataSourceConfig,
+  DataSourceInfo,
+} from "@connectors/types/data_source_config";
+
+import {
+  getDatabaseChildrenOfDocument,
+  getNotionPageFromConnectorsDb,
+  getPageChildrenOfDocument,
+} from "./connectors_db_helpers";
+
+/** Compute the parents field for a notion document See the [Design
+ * Doc](https://www.notion.so/dust-tt/Engineering-e0f834b5be5a43569baaf76e9c41adf2?p=3d26536a4e0a464eae0c3f8f27a7af97&pm=s)
+ * and the field documentation [in core]() for relevant details
+ */
+async function _getParents(
+  dataSourceInfo: DataSourceInfo,
+  document: {
+    notionId: string;
+    parentType: string | null | undefined;
+    parentId: string | null | undefined;
+  }
+): Promise<string[]> {
+  const parents: string[] = [document.notionId];
+  switch (document.parentType) {
+    case "workspace":
+      return parents;
+    case "block":
+      // rare cases in which doing something here is useful
+      // are ignored for now, see the design doc for details
+      return parents;
+    case "page":
+    case "database": {
+      // retrieve the parent from notion connectors db
+      // and add it to the parents array
+      const parent = await getNotionPageFromConnectorsDb(
+        dataSourceInfo,
+        document.parentId as string // (cannot be null here)
+      );
+      if (!parent) {
+        // The parent is either not synced yet (not an issue, see design doc) or
+        // is not in Dust's scope, in both cases we can just return the page id
+        return parents;
+      }
+      return parents.concat(
+        await getParents(dataSourceInfo, {
+          notionId: parent.notionPageId,
+          parentType: parent.parentType,
+          parentId: parent.parentId,
+        })
+      );
+    }
+    default:
+      throw new Error(`Unhandled parent type ${document.parentType}`);
+  }
+}
+
+export const getParents = memoize(_getParents, (dataSourceInfo, document) => {
+  return `${dataSourceInfo.dataSourceName}:${document.notionId}`;
```

Comment:
Shouldn't we memoize with the activity id as key as well? We don't want to memoize this across activities do we?

Reply:
Yes; I thought activities didn't share memory (different processes) but it's not obvious they don't actually, so thanks; better to be explicit. IMHO the memoization sits well at the "updateAllParents" level, so will change code accordingly